File: python/p04.py
#Q: Find the largest palindrome made from the product of two 3-digit numbers.
#A: 906609

def is_palindrome(t):
    a = list(str(t))
    b = list(a)
    a.reverse()
    if b == a:
        return True
    return False

a = 0
last_j = 0
for i in xrange(999, 99, -1):
    if i == last_j:
        break
    for j in xrange(999, 99, -1):
        temp = i*j
        if is_palindrome(temp) and temp > a:
            a = temp
            last_j = j
            break
print(str(a))

File: ocaml/p04.ml
(* Q: Find the largest palindrome made from the product of two 3-digit numbers.
 * A: 906609 *)

open Base

let is_pal n =
  let s = Int.to_string n in
  let reversed = String.fold s ~init:[] ~f:(fun acc c -> c :: acc)
                 |> String.of_char_list in
  String.equal s reversed

let rec max_pal i j cur_max =
  let new_i, new_j = match i with
    | 0 -> 999, j - 1
    | _ -> i - 1, j in
  let n = i * j in
  let new_max = if n > cur_max && is_pal n then n else cur_max in
  match j with
  | 0 -> cur_max
  | _ -> max_pal new_i new_j new_max;;

max_pal 999 999 0;;
